optimize compilation

vtjnash opened the pull request:
It seems like we weren't doing a great job with picking which methods to compile. This attempts to be much better at picking them. The new SSA IR compiler is also much more pessimistic in the runtime than the old compiler, which was better optimized to work efficiently. This helps optimize a few more of its code patterns and fixes a couple more missing method signature types.

vtjnash committed:
separate julia compile stages more

This should allow us to do more of the build in parallel,
and will cause us to now re-use `.ji` cache files between
the -debug and -release julia binaries.

vtjnash committed:
build: separate -debug and -release build timing

Producing `.o` from `.ji` files takes a massive amount of memory (currently),
so it's likely to lead to less memory thrashing and OOM-kills to do this step in serial

ViralBShah commented:
Review from nanosoldier? Is that the new way to do perf benchmarking?

JeffBezanson commented on base/compiler/ssair/passes.jl:
```diff
@@ -150,19 +153,25 @@ end
 """
-function walk_to_defs(compact, defssa, typeconstraint, visited_phinodes=Any[])
+function walk_to_defs(compact::IncrementalCompact, @nospecialize(defssa), @nospecialize(typeconstraint), visited_phinodes::Vector{Any}=Any[])
+    if !isa(defssa, Union{OldSSAValue, SSAValue, NewSSAValue}) || !isa(compact[defssa], PhiNode)
```
should use `AnySSAValue`

vtjnash committed:
precompile: change compilation heuristic

Avoid dependence on the current running state, but instead depend on the quality of the dispatch cache.

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

JeffBezanson commented:
Nice, this does seem to use less memory and CI seems to be quite happy with it so far.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27604#issuecomment-399520380) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/82049b44a6e1a6581e702b9f5c75d462c846cd93/2ca102c_vs_46a090b/report.md). cc @ararslan

ararslan commented:
Looks pretty good overall. Going to rerun the scalar benchmarks to check, since those were the ones that showed the most regressions: @nanosoldier `runbenchmarks("scalar", vs=":master")`

vtjnash commented:
Sounds good. There shouldn't be differences, so this run is just a sanity test, and a bit of a smoke screen test for nanosoldier itself.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27604#issuecomment-399616095) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/bed0e86a2d627577f2e18f119f18e384bcc9e2d6/4aa9796_vs_0eaa8b9/report.md). cc @ararslan
